yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+                eventPostFailures.incrementAndGet
+                logWarning(s"Connection exception submitting $entityDescription\n$e", e)
+                true
+
+              case e: RuntimeException =>
+                // this is probably a retry timeout event; some Hadoop versions don't
+                // rethrow the exception causing the problem, instead raising an RTE
+                eventPostFailures.incrementAndGet
+                logWarning(s"Runtime exception submitting $entityDescription\n$e", e)
+                // same policy as before: retry on these
+                true
+
+              case e: Exception =>
+                // something else has gone wrong.
+                eventPostFailures.incrementAndGet
+                logWarning(s"Could not handle history entity: $entityDescription\n$e", e)
+                false
+            }
+          }
+        }
+      }
+      logDebug(s"after pushEntities: ${entityList.size}")
+    }
+  }
+
+  /**
+   * Process the action placed in the queue
+   * @param action action
+   * @return true if the queue processor must now exit
+   */
+  private def processAction(action: QueuedAction): Boolean = {
+    action match {
+
+      case StopQueue() =>
+        logDebug("Stop queue action received")
+        true
+
+      case event: HandleSparkEvent =>
+        try {
+          handleEvent(event)
+          false
+        } catch {
+          case NonFatal(e) => false
+        }
+
+      case FlushTimelineEvents() =>
+        logDebug("Flush queue action received")
+        flushCurrentEventsToATS
+        false
+    }
+  }
+
+  /**
+   * If the event reaches the batch size or flush is true, push events to ATS.
+   *
+   * @param event event. If null no event is queued, but the post-queue flush logic
+   *              still applies
+   * @return true if the event should stop the service.
+   */
+  private def handleEvent(event: HandleSparkEvent): Unit = {
+    var push = false
+    // if we receive a new appStart event, we always push
+    // not much contention here, only happens when service is stopped
+    entityLock synchronized {
+      logDebug(s"Processing event $event")
+      if (eventsProcessed.get() % 1000 == 0) {
+        logDebug(s"${eventsProcessed} events are processed")
+      }
+      val sparkEvent = event.sparkEvent
+      sparkEvent match {
+        case start: SparkListenerApplicationStart =>
+          // we already have all information,
+          // flush it for old one to switch to new one
+          logDebug(s"Handling application start event: $event")
+          appName = start.appName
+          userName = start.sparkUser
+          startTime = start.time
+          if (startTime == 0) {
+            startTime = now()
+          }
+          setContextAppAndAttemptInfo(
+            start.appId,
+            start.appAttemptId)
+          // flush old entity
+          resetCurrentEntity
+          appStartEventProcessed.set(true)
```

Comment:
You can get events coming in before the app start, in particular {{onEnvironmentUpdate}} comes in. It's currently going out as an event using the entity built from the Yarn app id (which is guaranteed to be unique cluster-wide, and known from the moment the service starts). 

Some tests did re-use the queue, but they should all have been moved off it with the move to multi-attempts (and fixed entity events).

That's an interesting q. about event history without the full list of events ... I don't think there's any test for it.